### PR TITLE
refactor(stream): refactor state of topN executor to one state table

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Refactor internal state of topN executor to one state table.
The basic idea is to find the boundary keys at offset and offset + limit, and compare the boundary before and after modification. Then emit the differential result to downstream. The code is a kind of verbose, but it will be refined soon.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#3489

### CODE REVIEW

File: src/stream/src/executor/top_n_new.rs [hunk 1]
```diff
@@ -0,0 +1,904 @@
...
+                                if end_row.ordered_key != new_end_row.ordered_key {
+                                    new_ops.push(Op::Delete);
+                                    new_rows.push(end_row.row);
+                                }
+
+                                new_ops.push(Op::Insert);
+                                new_rows.push(row.clone());
+                            }
+                        }
+                    } else if self.offset + 1 == self.managed_state.total_count() {
```
Comment (BowenXiao1999) [inline src/stream/src/executor/top_n_new.rs:257]: Add some doc on the `if` like above?

File: src/stream/src/executor/top_n_new.rs [hunk 2]
```diff
@@ -0,0 +1,904 @@
...
+                            if end_row.ordered_key != new_end_row.ordered_key && end_row.is_valid()
+                            {
+                                assert!(new_end_row.ordered_key <= end_row.ordered_key);
+                                new_ops.push(Op::Delete);
+                                new_rows.push(end_row.row);
+                            }
+                        } else if !end_row.is_valid() {
+                            // no limit
+                            new_ops.push(Op::Insert);
+                            new_rows.push(row);
```
Comment (BowenXiao1999) [inline src/stream/src/executor/top_n_new.rs:287]: re-use a function on this?

File: src/stream/src/executor/top_n_new.rs [hunk 3]
```diff
@@ -0,0 +1,904 @@
...
+                        }
+                    } else if self.offset + 1 == self.managed_state.total_count() {
+                        // handle the boundary case
+                        let (new_start_row, _) = self
+                            .managed_state
+                            .find_range(self.offset, self.limit, epoch)
+                            .await?;
+                        new_ops.push(Op::Insert);
+                        new_rows.push(new_start_row.row);
+                    } else if let Some(start_key) = start_row.ordered_key.as_ref() {
```
Comment (BowenXiao1999) [inline src/stream/src/executor/top_n_new.rs:265]: ditto

File: src/stream/src/executor/managed_state/top_n/top_n_state_new.rs [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+            row: Row::default(),
+        };
+        let mut first_row = invalid_row.clone();
+        if let Some(next_res) = state_table_iter.next().await {
+            first_row = self.get_topn_row(next_res);
+        }
+
+        // fetch start row
+        let mut start_row = first_row;
+        if offset > 0 {
```
Comment (lmatz) [inline src/stream/src/executor/managed_state/top_n/top_n_state_new.rs:152]: Probably no need?

File: src/stream/src/executor/top_n_new.rs [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+            internal_key_indices.push(*pk_index);
+            internal_order_types.push(OrderType::Ascending);
+        }
+    }
+    let internal_data_types = internal_key_indices
+        .iter()
+        .map(|idx| schema.fields()[*idx].data_type())
+        .collect();
+    (
+        internal_key_indices,
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/stream/src/executor/top_n_new.rs:118]: This vector not only contains column indices of primary key but also the "order by" columns.
For example we have `table t1 (v1 int, v2 int)`, and with query `select * from t1 order by t1 limit 3`. The pk for this query is `row_id`, and the order column is v1 then this vector will contains indices to `v1` and `row_id`.

**Review by wcy-fdu — COMMENTED** [review]
Will we bench three table top_n and one table top_n?
I have bench cell_based encoding and row-based encoding on top_n, maybe the benchmark env can also used for this PR.

File: src/stream/src/executor/top_n_new.rs (hunk 2 above)
Comment (StrikeW) [reply src/stream/src/executor/top_n_new.rs:274]: Do you mean wrap those two lines to a function?  That will make the code cleaner, but I think current way is also readable. 
```
new_ops.push(Op::Insert);
new_rows.push(row);
```

**Comment by StrikeW** [comment]
> Will we bench three table top_n and one table top_n? I have bench cell_based encoding and row-based encoding on top_n, maybe the benchmark env can also used for this PR.

The top-N needs to be enhanced to two-phase top-N after this PR. So it is not stable yet, I think we should bench a more stable version.

**Review by lmatz — APPROVED** [review]

**Pull request merged**